worlds/blasphemous/Rules.py

```diff
@@ -1,1479 +1,3333 @@
+    # Items
+    set_rule(world.get_location("GotP: Lower west shaft", player),
+        lambda state: state.has("D02Z02S04[E]", player))
+    set_rule(world.get_location("GotP: Upper west shaft", player),
+        lambda state: state.has("D02Z02S04[NE]", player) or \
+            ((state.has("D02Z02S04[W]", player) or \
+                state.has("D02Z02S04[E]", player) and \
+                    state.has("Dash Ability", player)) and \
+                        (state.has("Purified Hand of the Nun", player) or \
+                            state.has("Wall Climb Ability", player))) or \
+                                (state.has("D02Z02S04[SE]", player) and \
+                                    (state.has("Wall Climb Ability", player) or \
+                                        state.has("Purified Hand of the Nun", player) and \
+                                            state._blasphemous_can_enemy_upslash(difficulty, enemy, player))))
+    set_rule(world.get_location("GotP: West shaft Child of Moonlight", player),
+        lambda state: (state.has("D02Z02S04[NE]", player) or \
+            state.has("D02Z02S04[W]", player) or \
+                state.has("D02Z02S04[E]", player) and \
+                    state.has("Dash Ability", player) or \
+                        state.has("D02Z02S04[SE]", player) and \
+                            (state.has("Wall Climb Ability", player) or \
+                                state.has("Purified Hand of the Nun", player) and \
+                                    state._blasphemous_can_enemy_upslash(difficulty, enemy, player))) and \
+                                        (state.has("Blood Perpetuated in Sand", player) and \
+                                            state.has("Dash Ability", player) or \
+                                                state.has("Purified Hand of the Nun", player) and \
+                                                    state._blasphemous_can_enemy_bounce(difficulty, enemy) or \
+                                                        state.has_any({"Lorquiana", "Cante Jondo of the Three Sisters", "Verdiales of the Forsaken Hamlet", "Cantina of the Blue Rose"}, player) or \
+                                                            state._blasphemous_aubade(player)) or \
+                                                                (state.has("D02Z02S04[NE]", player) or \
+                                                                    state.has("D02Z02S04[W]", player) or \
+                                                                        state.has("D02Z02S04[E]", player) and \
+                                                                            state.has("Dash Ability", player) or \
+                                                                                state.has("D02Z02S04[SE]", player)) and \
+                                                                                    state._blasphemous_pillar(player))
```

Comment:
YIKES

I know nothing about the game, so it's totally possible that this is necessary, but a lot of these conditions seem LONG, way longer than I'd expect them to be for a game making good use of regions.

If that's not the case and this one check's logic is just extremely complicated, then so be it :D

Reply:
Yeah it's pretty ugly but these really long conditions are (unfortunately) necessary because of plans to add Door Randomizer at some point in the future. And also to make the logic consistent with how it is written in standalone Blasphemous randomizer so it's easier to pick out any potential mistakes between the two, which already has Door Randomizer implemented.

So yeah... I would love for it not to be like that but it kind of needs to be, lol

Reply:
What do you think about something like this? Basically putting this into Disjunctive Normal Form, where every new "or" starts on a new line on the same indent level. Also, getting rid of the line continuations using \ and just wrapping the whole thing in () which I believe you can do

```
lambda state: (
                           (
                               state.has("D02Z02S04[NE]", player)
                               or state.has("D02Z02S04[W]", player)
                               or state.has("D02Z02S04[E]", player) and state.has("Dash Ability", player)
                               or state.has("D02Z02S04[SE]", player) and (
                                   state.has("Wall Climb Ability", player)
                                   or state.has("Purified Hand of the Nun", player) and state._blasphemous_can_enemy_upslash(difficulty, enemy, player)
                               )
                            ) 
                            and (
                                state.has("Blood Perpetuated in Sand", player) and state.has("Dash Ability", player)
                                or state.has("Purified Hand of the Nun", player) and state._blasphemous_can_enemy_bounce(difficulty, enemy)
                                or state.has_any({"Lorquiana", "Cante Jondo of the Three Sisters", "Verdiales of the Forsaken Hamlet", "Cantina of the Blue Rose"}, player) 
                                or state._blasphemous_aubade(player)
                            ) 
                            or state.has("D02Z02S04[NE]"
                            or state.has("D02Z02S04[W]", player)
                            or state.has("D02Z02S04[E]", player) and state.has("Dash Ability", player) 
                            or state.has("D02Z02S04[SE]", player)) and state._blasphemous_pillar(player)
                      )
 ```

Reply:
I don't know if I actually converted it right, but I hope I got the point across. :D

Reply:
Here's a similar example from The Witness that works 100%, although not in the context of a lambda expression:

```
return (
    self.can_reach("Keep", "Region", player)
    and self.can_reach("Keep 4th Pressure Plate to Keep Tower", "Entrance", player)
    and (
        self.can_reach("Keep Tower to Keep", "Entrance", player)
        or hedge_access
    )
)
```